src/hotspot/share/gc/parallel/mutableSpace.cpp

```diff
@@ -230,6 +230,24 @@ void MutableSpace::object_iterate(ObjectClosure* cl) {
   }
 }
 
+HeapWord* MutableSpace::block_start(const void* addr) const {
+  if (addr >= top()) {
+    return nullptr;
+  }
+
+  HeapWord* cur_addr = bottom();
+  while (cur_addr <= addr) {
```

Comment:
As already described in https://github.com/openjdk/jdk/pull/21870#issuecomment-2454964142, this will not work in the general case, if the heap is not walkable. In a debug build you'll run into the assertion once you arrive in the unallcoated TLAB area (which you don't want during error reporting).
Even worse, in the product build you can crash or run this loop infinitely, depending on what data `obj->size()` will fnd in the unallocated TLAB space.

Reply:
What would be nice would be something like `oopDesc::safe_klass_or_null()` or similar, feeding into a corresponding `oopDesc::size_given_klass_safe_or_0()`. The former would check the klass word for validity before dereferencing - `CompressedKlassPointers::is_encodable(p)` and then the load of layouthelper etc should happen with SafeFetch. Alternatively (and a bit more unsafe), check the readability of Klass* with SafeFetch beforehand, then call normal size_given_klass.

Reply:
> > this will not work in the general case, if the heap is not walkable.
> 
> True, but this is the best-effort approach used in other GCs, as far as I can tell. Is there a real use case that warrants a more sophisticated variant?

The **only** use case for this code during hs_err reporting for heap-addresses not pointing at the beginning of an oop. I think we should be conservative here, because a secondary crash will cut the information available in the hs_err file and will therefor do more harm then being helpful.

Reply:
> What would be nice would be something like `oopDesc::safe_klass_or_null()` or similar, feeding into a corresponding `oopDesc::size_given_klass_safe_or_0()`. The former would check the klass word for validity before dereferencing - `CompressedKlassPointers::is_encodable(p)` and then the load of layouthelper etc should happen with SafeFetch. Alternatively (and a bit more unsafe), check the readability of Klass* with SafeFetch beforehand, then call normal size_given_klass.

We already have [LocationPrinter::is_valid_obj()](https://github.com/openjdk/jdk/blob/ac82a8f89c7066fb1d379b12bcfd68053cb39ba4/src/hotspot/share/gc/shared/locationPrinter.cpp#L33) which uses [Klass::is_valid()](https://github.com/openjdk/jdk/blob/ac82a8f89c7066fb1d379b12bcfd68053cb39ba4/src/hotspot/share/oops/klass.cpp#L1038) to check the validity of an oop. I don't think we need `SafeFetch` here.